hummingbot/core/data_type/trade_fee.py

```diff
@@ -1,66 +1,37 @@
 import typing
+from abc import ABC, abstractmethod
 from dataclasses import dataclass, field
 from decimal import Decimal
-from enum import Enum
 from typing import Dict, List, Optional, Tuple
 
-from hummingbot.connector.utils import split_hb_trading_pair, combine_to_hb_trading_pair
+from hummingbot.connector.utils import combine_to_hb_trading_pair, split_hb_trading_pair
 from hummingbot.core.event.utils import interchangeable
 
 if typing.TYPE_CHECKING:  # avoid circular import problems
     from hummingbot.connector.exchange_base import ExchangeBase
-
-
-class TradeFeePercentageApplication(Enum):
-    """
-    Indicates how the trade fee object must be applied.
-
-    `AddedToCost` is used for normal buy orders and third-token percentage fees for both buy and sell orders.
-    `DeductedFromReturns` is used for normal sell orders and Binance-style buy orders, where an order to buy
-    1000 USDT worth of BTC with a 0.1% fee will return 999 USDT worth of BTC to the user.
-    """
-    AddedToCost = 0
-    DeductedFromReturns = 1
+    from hummingbot.core.data_type.order_candidate import OrderCandidate
 
 
 @dataclass
-class TradeFee:
+class TradeFeeBase(ABC):
     """
     Contains the necessary information to apply the trade fee to a particular order.
     """
     percent: Optional[Decimal] = None
     percent_token: Optional[str] = None  # only set when fee charged in third token (the Binance BNB case)
-    percentage_application: TradeFeePercentageApplication = TradeFeePercentageApplication.AddedToCost
-    flat_fees: List[Tuple[str, Decimal]] = field(default_factory=tuple)  # list of (asset, amount) tuples
-
-    @classmethod
-    def from_json(cls, data: Dict[str, any]) -> "TradeFee":
-        percent = data["percent"]
-        percent_token = data.get("percent_token")
-        percentage_application = data.get("percentage_application")
-        percentage_application = (
-            TradeFeePercentageApplication(int(percentage_application))
-            if percentage_application is not None
-            else TradeFeePercentageApplication.AddedToCost
-        )
-        flat_fees = [
-            (str(fee_entry["asset"]), Decimal(fee_entry["amount"]))
-            for fee_entry in data["flat_fees"]
-        ]
-        return TradeFee(percent, percent_token, percentage_application, tuple(flat_fees))
+    flat_fees: List[Tuple[str, Decimal]] = field(default_factory=list)  # list of (asset, amount) tuples
 
     def to_json(self) -> Dict[str, any]:
         return {
-            "percent": float(self.percent),
+            "percent": None if self.percent is None else float(self.percent),
```

Comment:
Does it work correctly with None? I imagined it would be 0 when it does not apply.

Reply:
Agreed, better to express it as zero instead of None. Addressed [here](https://github.com/CoinAlpha/hummingbot/pull/21/commits/2f2af2f6a8e79047586300e308ec528d5efe5ba8)